baseplate/clients/redis_cluster.py

```diff
@@ -0,0 +1,225 @@
+from baseplate.lib import metrics
+
+
+# We want to be able to combine blocking behaviour with the ability to read from replicas
+# Unfortunately this is not provide as-is so we cmobine two connection pool classes to provide
+# the desired behaviour.
+class ClusterWithReadReplicasBlockingConnectionPool(rediscluster.ClusterBlockingConnectionPool):
+    # pylint: disable=arguments-differ
+    def get_node_by_slot(self, slot: int, read_command: bool = False) -> Dict[str, Any]:
+        """
+        Get a random node from the slot, including master
+        """
+        nodes_in_slot = self.nodes.slots[slot]
+        if read_command:
+            random_index = random.randrange(1, len(nodes_in_slot))
+            return nodes_in_slot[random_index]
+
+        return nodes_in_slot[0]
+
+
+def cluster_pool_from_config(
+    app_config: config.RawConfig, prefix: str = "rediscluster.", **kwargs: Any
+) -> rediscluster.ClusterConnectionPool:
+    """Make a ClusterConnectionPool from a configuration dictionary.
+
+    The keys useful to :py:func:`clusterpool_from_config` should be prefixed, e.g.
+    ``rediscluster.url``, ``rediscluster.max_connections``, etc. The ``prefix`` argument
+    specifies the prefix used to filter keys.  Each key is mapped to a
+    corresponding keyword argument on the :py:class:`rediscluster.ClusterConnectionPool`
+    constructor.
+
+    Supported keys:
+
+    * ``url`` (required): a URL like ``redis://localhost/0``.
+    * ``max_connections``: an integer maximum number of connections in the pool
+    * ``skip_full_coverage_check``: Skips the check of cluster-require-full-coverage
+      config, useful for clusters without the CONFIG command (like aws)
+    * ``nodemanager_follow_cluster``: Tell the node manager to reuse the last set of
+      nodes it was operating on when intializing.
+    * ``read_from_replicas``: (Boolean) Whether the client should send all read queries to
+        replicas instead of the primary
+    * ``timeout``: how long to wait for sockets to connect. e.g.
+        ``200 milliseconds`` (:py:func:`~baseplate.lib.config.Timespan`)
+    """
+
+    assert prefix.endswith(".")
+
+    parser = config.SpecParser(
+        {
+            "url": config.String,
+            "max_connections": config.Optional(config.Integer, default=None),
+            "timeout": config.Optional(config.Timespan, default=100),
+            "read_from_replicas": config.Optional(config.Boolean, default=False),
+            "skip_full_coverage_check": config.Optional(config.Boolean, default=True),
+            "nodemanager_follow_cluster": config.Optional(config.Boolean, default=None),
+            "decode_responses": config.Optional(config.Boolean, default=True),
+        }
+    )
+
+    options = parser.parse(prefix[:-1], app_config)
+
+    # We're explicitly setting a default here because of https://github.com/Grokzen/redis-py-cluster/issues/435
+    kwargs.setdefault("max_connections", options.max_connections)
```

Comment:
yikes! thanks for catching this and documenting it!

Reply:
but it looks like the default in our config is also `None` so would we just be causing the same bug?

Reply:
oh good catch! I think I unset the default for some tests and forgot to change it back, I'll fix it